### PR TITLE
👩‍💼👨‍💼 feat: pm role

### PR DESCRIPTION
**Pull request by florian-hoenicke**
It turned out to be difficult for the users to specify the tasks in the exact same format required by us.
This pr aims to solve this issue by asking for the required information.
It always takes a random name from our product people leon, susana saahil

[Watch the video](https://user-images.githubusercontent.com/11627845/234832586-7765cd5d-9cdc-49ba-9bd4-5703eb40b25d.mov)

### CODE REVIEW

File: src/apis/gpt.py [hunk 1]
```diff
@@ -96,29 +96,39 @@ def on_llm_new_token(self, token: str, **kwargs: Any) -> None:
...
+        self.print_stream = print_stream
+        self.print_costs = print_costs
+        for message in messages:
+            if os.environ['VERBOSE'].lower() == 'true':
+                if isinstance(message, SystemMessage):
+                    print_colored('system - prompt', message.content, 'magenta')
+                elif isinstance(message, HumanMessage):
+                    print_colored('user - prompt', message.content, 'blue')
+                elif isinstance(message, AIMessage):
+                    print_colored('assistant - prompt', message.content, 'green')
```
Comment (joschkabraun) [inline src/apis/gpt.py:118]: maybe `assistant - response` is better?

File: src/options/generate/generator.py [hunk 2]
```diff
@@ -320,3 +347,87 @@ def summarize_error(self, error):
         error_summary = conversation.chat(template_summarize_error.format(error=error))
         return error_summary
 
+    def refine_specification(self):
+        pm = get_random_employee('pm')
+        print(f'{pm.emoji}👋 Hi, I\'m {pm.name}, a PM at Jina AI. Gathering the requirements for our engineers.')
+        self.refine_task(pm)
+        self.refine_test(pm)
+        print(f'''
```
Comment (joschkabraun) [inline src/options/generate/generator.py:355]: Should we ask for confirmation in the end, to confirm description & test?

File: src/options/generate/ui.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+product_manager_names = [
+    ('Leon', 'm'),
+    ('Saahil', 'm',),
+    ('Susana', 'f')
+]
+engineer_names = [
+    ('Joschka', 'm'),
+    ('Johannes', 'm'),
+    ('Johanna', 'f'),
```
Comment (joschkabraun) [inline src/options/generate/ui.py:12]: we have biased sampling but I don't mind 😅

File: src/options/generate/templates_system.py [hunk 4]
```diff
@@ -107,3 +107,153 @@ def foo(self, docs: DocumentArray, **kwargs) => DocumentArray:
...
+Example for the description "parser":
+input: ❌
+output: ❌
+credentials: n/a
+database access: n/a
+
+**prompt.txt**
+```text
+Please provide the input and output format.
+```
```
Comment (joschkabraun) [inline src/options/generate/templates_system.py:183]: we can drop trailing white spaces

File: src/options/generate/templates_system.py [hunk 5]
```diff
@@ -107,3 +107,153 @@ def foo(self, docs: DocumentArray, **kwargs) => DocumentArray:
...
+The microservice takes nothing as input and returns the current time.
+```
+PM:
+**test-final.txt**
+```text
+The test takes nothing as input and asserts that the output is a string.
+```
+
+
+'''
```
Comment (joschkabraun) [inline src/options/generate/templates_system.py:259]: we can drop trailing white spaces

File: src/options/generate/generator.py [hunk 6]
```diff
@@ -320,3 +347,87 @@ def summarize_error(self, error):
...
+            if task_final:
+                self.microservice_specification.task = task_final
+                break
+            if question:
+                task_description = self.get_user_input(pm, question)
+                messages.extend([HumanMessage(content=task_description)])
+            else:
+                task_description = self.get_user_input(pm, agent_response_raw + '\n: ')
+
+    def refine_test(self, pm):
```
Comment (joschkabraun) [inline src/options/generate/generator.py:387]: there is code duplication in `refine_task` and `refine_test`

File: src/apis/gpt.py (hunk 1 above)
Comment (florian-hoenicke) [reply src/apis/gpt.py:120]: I wanted to illustrate that this log output is part of the prompt we give to gpt.
Now we can differentiate between messages that are part of the prompt and generated messages.
It is new that we also have ai messages handed over as prompt

File: src/options/generate/generator.py (hunk 2 above)
Comment (florian-hoenicke) [reply src/options/generate/generator.py:426]: Good idea. We could think about having this in the future.
Let's have a discussion about pros and cons.

File: src/options/generate/ui.py (hunk 3 above)
Comment (florian-hoenicke) [reply src/options/generate/ui.py:12]: nice catch!
I used the three examples to let gpt4 generate the rest.
I removed them now

File: src/options/generate/templates_system.py (hunk 4 above)
Comment (florian-hoenicke) [reply src/options/generate/templates_system.py:183]: done

File: src/options/generate/templates_system.py (hunk 5 above)
Comment (florian-hoenicke) [reply src/options/generate/templates_system.py:260]: done

File: src/options/generate/generator.py (hunk 6 above)
Comment (florian-hoenicke) [reply src/options/generate/generator.py]: done

File: src/apis/gpt.py (hunk 1 above)
Comment (joschkabraun) [reply src/apis/gpt.py:120]: Ah yes makes sense! Now I understand

**Pull request merged**